src/pip/_internal/index.py

```diff
@@ -802,6 +804,11 @@ def get_page(cls, link, skip_archives=True, session=None):
                 url = urllib_parse.urljoin(url, 'index.html')
                 logger.debug(' file: URL is directory, getting %s', url)
 
+            if blacklisted_hosts and scheme != 'file':
```

Comment:
i believe `if scheme != 'file' and netloc in blacklisted_hosts` might be a sufficient check not needing any nesting